proposals/infra/INF-0004-binding-resource-binding-diagnostics.md

```diff
@@ -0,0 +1,212 @@
+valid binding prefixes for that resource, and what are all the possible 
+diagnostics that can be emitted if the resource is given a binding 
+prefix that isn't in the set of valid binding prefixes?
+The design below aims to specify the answer to these questions.
+
+## Proposed solution
+Firstly, the most common case is when a resource type that is not a 
+user-defined type is bound as a resource. Any resource will have an 
+associated ResourceAttr attribute, from which we can determine the 
+resource class and resource kind. These two inputs are sufficient to
+determine the expected binding prefix, and any other binding prefix
+will result in an error. The recommended binding prefix will be
+suggested in the diagnostic. The table below specifies what binding 
+prefix will be expected for any resource type that falls under the
+specified resource class. 
+
+| Resource Class | Binding Prefix | Diagnostic |
+|-|-|-|
+| Sampler | s | "Object type '%0' with resource class "Sampler" expects binding prefix 's', but binding prefix '%1' was given." |
+| SRV | t | "Object type '%0' with resource class "SRV" expects binding prefix 't', but binding prefix '%1' was given." |
+| UAV | u | "Object type '%0' with resource class "UAV" expects binding prefix 'u', but binding prefix '%1' was given." |
+| CBuffer | b |  "Object type '%0' with resource class "CBuffer" expects binding prefix 'b', but binding prefix '%1' was given." |
+
+There are certain resource kinds within the SRV resource
+class that are "sampleable". These resources are typically
+prefixed with "Texture" in actual HLSL source. So, there will
+be a special check to see whether the candidate resource is 
+prefixed with "Texture", and if so, the "s" binding prefix will
+also be permitted for that resource.
+
+If the given candidate resource is a user-defined type (UDT), then further
+analysis is necessary. The first step is to gather all registers that
+are being bound to this declaration, and collect the binding prefixes
+that are being specified. The UDT must have at least one valid resource
+that can be bound to the provided binding prefix. If not, an error must be
+emitted stating that "No resource contained in struct '%0' can be bound
+to binding prefix '%1'". There are no issues if a UDT has more resources
+than there are register binding statements, the resources will be bound to
+the next available space automatically, and so compilation can succeed.
+Additionally, a singular resource can be bound multiple times to different
+spaces, so multiple binding prefixes are allowed for a single resource.
+Below is a table that represents the different behaviors that could arise
+given some examples of different UDT's:
+
+| UDT | Binding statements | Diagnostic |
+|-|-|-|
+| struct Foo {<br>  float f;<br>  Buffer<float> Buf;<br>  RWBuffer<float> RWBuf;<br>};| register(t0) : register(s0); | None. Success because Buf gets bound to t0 and RWBuf gets bound to s0, and f is skipped. 

```

Comment:
I think that this is saying you can write things like this?

```
struct Foo { 
  float f; 
  Buffer<float> Buf; 
  RWBuffer<float> RWBuf; 
};

Foo Bar : register(t0):register(s0);
```

TIL, I'd never seen that before.  I think we need to do some work to make clang parse this:

```
<source>:2:23: error: expected ';' after top level declarator
    2 | Foo Bar : register(t0):register(s0);
      |                       ^
      |                       ;
```
```

Reply:
Incidentally, as this is pretty unusual syntax, I'd prefer to have more complex code examples to read.  Splitting between the UDT and "binding statements" columns means that I'm not entirely sure what syntax is intended here.

Reply:
Removed the binding statement column, just made a singular code column along with the expected diagnostic